pkg/cmd/pulumi/state_edit_frontend.go

```diff
@@ -0,0 +1,238 @@
+// Copyright 2016-2023, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sort"
+
+	"github.com/pulumi/pulumi/pkg/v3/resource/deploy"
+	"github.com/pulumi/pulumi/pkg/v3/resource/stack"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/apitype"
+	yaml "gopkg.in/yaml.v3"
+)
+
+type snapshotText string
+
+// snapshotFrontend is an interface that enables roundtripping state to a text format for editing.
+type snapshotFrontend interface {
+	SnapshotToText(deploy.Snapshot) (snapshotText, error)
```

Comment:
A round-tripping interface like this is well-suited to fuzz testing.
Consider writing a fuzz test or a [rapid](https://pkg.go.dev/pgregory.net/rapid)-based test (which can also be turned into a testing.F fuzz test with [MakeFuzz](https://pkg.go.dev/pgregory.net/rapid#MakeFuzz)) to ensure that this never corrupts a snapshot.